src/mission_companion.cpp

```diff
@@ -1173,6 +1339,350 @@ bool talk_function::labor_return( npc &p )
+    } else {
+        itemlist = "forest" ;
+    }
+    std::string output = "Notes: \nSend a companion to gather materials for the next camp upgrade.\n \n"
+        "Skill used: survival\n"
+        "Difficulty: N/A \n"
+        "Gathering Possibilities:\n";
+
+    // Functions like the debug item group tester but only rolls 6 times so the player doesn't have perfect knowledge
+    std::map<std::string, int> itemnames;
+    for (size_t a = 0; a < 6; a++) {
+        const auto items = item_group::items_from( itemlist, calendar::turn );
+        for( auto &it : items ) {
+            itemnames[it.display_name()]++;
+        }
+    }
+    // Invert the map to get sorting!
+    std::multimap<int, std::string> itemnames2;
+    for (const auto &e : itemnames) {
+        itemnames2.insert(std::pair<int, std::string>(e.second, e.first));
+    }
+    for (const auto &e : itemnames2) {
+        output = output + "> " + e.second + "\n";
+    }
+    std::vector<std::shared_ptr<npc>> npc_list = companion_list( p, "_faction_camp_gathering" );
+    output = output + " \nRisk: Very Low\n"
+            "Time: 3 Hours, Repeated\n"
+            "Positions: "+ to_string(npc_list.size()) +"/3\n";
+    return output;
+}
+
+std::string talk_function::om_next_upgrade( std::string bldg ){
+    std::string comp = "";
+    int phase = bldg.find_last_of("_");
+    comp = bldg.substr(phase+1);
+    int value = atoi(comp.c_str()) + 1;
+    comp = bldg.substr(0,phase+1) + to_string(value);
+
+    //This needs to be generic
+    if (comp == "faction_base_camp_19"){
+        return "null";
+    }
+    return comp;
+}
+
+bool talk_function::om_min_level( std::string target, std::string bldg ){
+    std::string comp = "";
+    int phase_target = target.find_last_of("_");
+    int phase_bldg = bldg.find_last_of("_");
+    //comparing two different buildings
+    if( target.substr(0,phase_target+1) != bldg.substr(0,phase_bldg+1)){
+        return false;
+    }
+    //not high enough level
+    if( atoi(target.substr(phase_target+1).c_str()) > atoi(bldg.substr(phase_bldg+1).c_str()) ) {
+        return false;
+    }
+    return true;
+}
+
+bool talk_function::upgrade_return( npc &p )
+{
+    npc *comp = companion_choose_return( p.name + "_faction_upgrade_camp", calendar::turn - 3_hours );
+    if (comp == NULL){
+        return false;
+    }
+
+    int exp = 3;
+    int y,i = 0;
+    while (i < exp){
+        y = rng( 0, 100 );
+        if (y < 80){
+            comp->practice( skill_fabrication, 10);
+        } else if (y < 90){
+            comp->practice( skill_survival, 5);
+        } else {
+            comp->practice( skill_mechanics, 5);
+        }
+        i++;
+    }
+
+    popup(_("%s returns from upgrading the camp having earned a bit of experience..."), comp->name.c_str());
+    companion_return( *comp );
+    return true;
+}
+
+bool talk_function::camp_gathering_return( npc &p, std::string task )
+{
+    //minimum working time is 3 hour
```

Comment:
Why are you repeating this over and over? And why is it *not* in the function `talk_function::upgrade_return`? Why is it there anyway - the function call already contains the 3 hours information?